node/modules/eth.go

```diff
@@ -38,47 +98,76 @@ type EthTransactionParams struct {
 	Indexer           index.Indexer
 }
 
-func MakeEthTransaction(cfg config.FevmConfig) func(EthTransactionParams) (eth.EthTransactionAPI, error) {
-	return func(params EthTransactionParams) (eth.EthTransactionAPI, error) {
-		// Prime the tipset cache with the entire chain to make sure tx and block lookups are fast
-		params.Lifecycle.Append(fx.Hook{
-			OnStart: func(context.Context) error {
-				go func() {
-					start := time.Now()
-					log.Infoln("Start prefilling GetTipsetByHeight cache")
-					_, err := params.ChainStore.GetTipsetByHeight(params.MetricsCtx, abi.ChainEpoch(0), params.ChainStore.GetHeaviestTipSet(), false)
-					if err != nil {
-						log.Warnf("error when prefilling GetTipsetByHeight cache: %w", err)
-					}
-					log.Infof("Prefilling GetTipsetByHeight done in %s", time.Since(start))
-				}()
-				return nil
-			},
-		})
+func MakeEthTransactionV1(cfg config.FevmConfig) func(EthTransactionParams, full.EthTipSetResolverV1) (full.EthTransactionAPIV1, error) {
+	return func(params EthTransactionParams, tipSetResolver full.EthTipSetResolverV1) (full.EthTransactionAPIV1, error) {
+		return makeEthTransaction(params, tipSetResolver, cfg.EthBlkCacheSize)
+	}
+}
 
-		return eth.NewEthTransactionAPI(
-			params.ChainStore,
-			params.StateManager,
-			params.StateAPI,
-			params.MpoolAPI,
-			params.Indexer,
-			params.EthEventsExtended,
-			cfg.EthBlkCacheSize,
-		)
+func MakeEthTransactionV2(cfg config.FevmConfig) func(EthTransactionParams, full.EthTipSetResolverV2) (full.EthTransactionAPIV2, error) {
+	return func(params EthTransactionParams, tipSetResolver full.EthTipSetResolverV2) (full.EthTransactionAPIV2, error) {
+		return makeEthTransaction(params, tipSetResolver, cfg.EthBlkCacheSize)
 	}
 }
 
-func MakeEthTrace(cfg config.FevmConfig) func(
+func makeEthTransaction(params EthTransactionParams, tipSetResolver eth.TipSetResolver, ethBlkCacheSize int) (full.EthTransactionAPIV1, error) {
+	// Prime the tipset cache with the entire chain to make sure tx and block lookups are fast
+	params.Lifecycle.Append(fx.Hook{
+		OnStart: func(context.Context) error {
+			go func() {
+				start := time.Now()
+				log.Infoln("Start prefilling GetTipsetByHeight cache")
+				_, err := params.ChainStore.GetTipsetByHeight(params.MetricsCtx, abi.ChainEpoch(0), params.ChainStore.GetHeaviestTipSet(), false)
+				if err != nil {
+					log.Warnf("error when prefilling GetTipsetByHeight cache: %w", err)
+				}
+				log.Infof("Prefilling GetTipsetByHeight done in %s", time.Since(start))
+			}()
+			return nil
```

Comment:
Can we pass in a context that gets canceled by `OnStop`? I guess this is existing code...

Reply:
yeah, we have `LifecycleCtx` for this, fixed